File: backend/app/schemas/quiz.py
from typing import Any, Literal

from pydantic import BaseModel, Field


class AnswerPayload(BaseModel):
    id: str
    text: str
    correct: bool = False


class QuestionPayload(BaseModel):
    id: str
    title: str
    type: Literal['text', 'image'] = 'text'
    image: str | None = ''
    imageCaption: str | None = ''
    multi: bool = False
    timeLimit: int = Field(default=45, ge=5)
    body: str | None = ''
    answers: list[AnswerPayload]


class QuizPayload(BaseModel):
    id: str | None = None
    ownerId: str
    title: str
    category: str | None = 'Общее'
    timeLimit: int | None = Field(default=45, ge=5)
    rules: str | None = ''
    questions: list[QuestionPayload]


class QuizOut(BaseModel):
    id: str
    ownerId: str
    title: str
    category: str
    timeLimit: int
    rules: str
    questions: list[dict[str, Any]]


class QuizListResponse(BaseModel):
    quizzes: list[QuizOut]


class QuizResponse(BaseModel):
    quiz: QuizOut

File: src/shared/lib/quiz.js
export function normalizeCode(code) {
  return String(code || '').trim().toUpperCase();
}

export function emptyQuestion() {
  return {
    id: crypto.randomUUID(),
    title: '',
    type: 'text',
    image: '',
    imageCaption: '',
    multi: false,
    timeLimit: 45,
    answers: [
      { id: crypto.randomUUID(), text: '', correct: true },
      { id: crypto.randomUUID(), text: '', correct: false },
      { id: crypto.randomUUID(), text: '', correct: false },
      { id: crypto.randomUUID(), text: '', correct: false }
    ]
  };
}

export function validateQuestion(question) {
  if (!String(question?.title || '').trim()) return 'Введите название вопроса.';
  if (question.type === 'text' && !String(question.body || '').trim()) return 'Введите текст вопроса.';
  if (question.type === 'image' && !String(question.image || '').trim()) return 'Загрузите изображение для вопроса.';
  if (!Array.isArray(question.answers) || question.answers.length < 2) return 'Добавьте минимум два ответа.';
  if (question.answers.some((answer) => !String(answer.text || '').trim())) return 'Заполните все варианты ответа.';
  const correctCount = question.answers.filter((answer) => answer.correct).length;
  if (!correctCount) return 'Отметьте правильный ответ.';
  if (!question.multi && correctCount !== 1) return 'Для вопроса с одним ответом выберите ровно один правильный вариант.';
  if (!Number(question.timeLimit) || Number(question.timeLimit) < 5) return 'Время на ответ должно быть не меньше 5 секунд.';
  return '';
}

export function validateQuiz(quiz) {
  const title = String(quiz.title || '').trim();
  if (!title) return 'Введите название викторины.';
  if (!quiz.questions.length) return 'Добавьте хотя бы один вопрос.';
  const questionError = quiz.questions.map(validateQuestion).find(Boolean);
  if (questionError) return questionError;
  return '';
}
